### PR TITLE
Fix LayoutBinding under AndroidX.

### PR DESCRIPTION
**Pull request by dellis1972**
Fixes: https://github.com/xamarin/xamarin-android/issues/6893

Context: 92cd68c7beb3a26dab55957b2e061faf6cdc1d89
Context: https://github.com/xamarin/AndroidX/blob/628cf049e0ef61e140cb17600df16ed801449814/source/AndroidXTargets.cshtml#L36

When:

 1. The [`$(AndroidFragmentType)`][0] MSBuild property is set, *and*
 2. The [`$(AndroidGenerateLayoutBindings)`][1]=True, *and*
 3. There is a Layout XML file which has an `id`, *and*
 4. Define `__HAVE_ANDROIDX__` C# symbol.

then [`LayoutBinding.cs`][2] (92cd68c7) fails to compile.

Repro:

 1. `dotnet new android`
 2. Edit `Resources/layout/activity_main.xml` so that the
    `<TextView/>` has `android:id="@+id/text_view"`:

        <TextView
          android:id="@+id/text_view"
          android:layout_width="wrap_content"                                     
          android:layout_height="wrap_content"                                    
          android:layout_centerInParent="true"                                    
          android:text="@string/app_text"                                         
        />

 3. Build!

        dotnet build -p:AndroidFragmentType=AndroidX.Fragment.App.Fragment \
          -p:AndroidGenerateLayoutBindings=True \
          -p:DefineConstants=__HAVE_ANDROIDX__

Expected results: no errors.

Actual results:

	…/LayoutBinding.cs(91,56): error CS0400: The type or namespace name 'Androidx' could not be found in the global namespace
	…/LayoutBinding.cs(91,135): error CS0400: The type or namespace name 'Androidx' could not be found in the global namespace

Update `LayoutBinding.cs` so that it uses the correct `AndroidX`
namespace.  Additionally, `FragmentManager` has been deprecated;
update `LayoutBinding.cs` to use `SupportFragmentManager`, which is
available on the `AndroidX.Fragment.App.FragmentActivity` class.

Finally, remove the use of a `static` field to share object state between
tasks. We have been on `msbuild` rather than `xbuild` on *nix for a while
now, so this workaround is no longer needed.

TODO: Add [`LayoutCodeBehind.md` docs][3] to
[MicrosoftDocs/xamarin-docs][4].

[0]: https://docs.microsoft.com/en-us/xamarin/android/deploy-test/building-apps/build-items#androidfragmenttype
[1]: https://docs.microsoft.com/en-us/xamarin/android/deploy-test/building-apps/build-properties#androidgeneratelayoutbindings
[2]: https://github.com/xamarin/xamarin-android/blob/16bbf6f0b0cd49f96f73197d28e8faf31aa9e499/src/Xamarin.Android.Build.Tasks/Resources/LayoutBinding.cs
[3]: https://github.com/xamarin/xamarin-android/blob/16bbf6f0b0cd49f96f73197d28e8faf31aa9e499/Documentation/guides/LayoutCodeBehind.md
[4]: https://github.com/MicrosoftDocs/xamarin-docs/tree/live/docs/android/deploy-test/building-apps

### CODE REVIEW

File: src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/Sdk/AutoImport.props [hunk 1]
```diff
@@ -29,8 +29,10 @@ https://github.com/dotnet/designs/blob/4703666296f5e59964961464c25807c727282cae/
   <ItemGroup Condition=" '$(EnableDefaultAndroidItems)' == 'true' ">
     <!-- Default Resource file inclusion -->
     <!-- https://developer.android.com/guide/topics/resources/providing-resources -->
-    <AndroidResource Include="$(MonoAndroidResourcePrefix)\*\*.xml" />
-    <AndroidResource Include="$(MonoAndroidResourcePrefix)\*\*.axml" />
+    <AndroidBoundLayout Include="$(MonoAndroidResourcePrefix)\layout*\*.xml" Condition=" '$(AndroidGenerateLayoutBindings)' == 'True' " />
```
Comment (dellis1972) [inline src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/Sdk/AutoImport.props:32]: Looking at the code I don't think we need this really. I'll be removing it

**Review by grendello — APPROVED** [review]

**Review by jonathanpeppers — APPROVED** [review]
LGTM

**Comment by jonpryor** [comment]
@dellis1972: can you please add a unit test for this?  Clearly we missed this for *years* -- this was added in 2018, via 92cd68c7beb3a26dab55957b2e061faf6cdc1d89 -- yet nobody reported it until 2022?

**Comment by jonpryor** [comment]
`Documentation/guides/LayoutCodeBehind.md` should also be updated: https://github.com/xamarin/xamarin-android/blob/782dd67512687c00b8379d8cc53c7c58a4d9795a/Documentation/guides/LayoutCodeBehind.md?plain=1#L537

(among other locations in that file)

**Comment by jonpryor** [comment]
Related: https://github.com/xamarin/AndroidX/blob/628cf049e0ef61e140cb17600df16ed801449814/source/AndroidXTargets.cshtml#L36

**Comment by jonpryor** [comment]
Related: the docs for `$(AndroidFragmentType)`: https://github.com/xamarin/xamarin-android/blob/main/Documentation/guides/building-apps/build-items.md?plain=1#L81-L85

should be moved into `build-properties.md`; it's *not* an item group.

**Comment by jonpryor** [comment]
Related: we should make `LayoutBinding.cs` nullable-reference-type aware; as-is, if `dotnet build` is used, using codebehind files results in lots of NRT-related warnings.

**Comment by jonpryor** [comment]
WIP commit message:

```markdown
Fixes: https://github.com/xamarin/xamarin-android/issues/6893

Context: 92cd68c7beb3a26dab55957b2e061faf6cdc1d89
Context: https://github.com/xamarin/AndroidX/blob/628cf049e0ef61e140cb17600df16ed801449814/source/AndroidXTargets.cshtml#L36

When:

 1. The [`$(AndroidFragmentType)`][0] MSBuild property is set, *and*
 2. The [`$(AndroidGenerateLayoutBindings)`][1]=True, *and*
 3. There is a Layout XML file which has an `id`, *and*
 4. (???Somehow???) The `__HAVE_ANDROIDX__` is defined

then `LayoutBinding.cs` (92cd68c7) fails to compile.

Repro:

 1. `dotnet new android`
 2. Edit `Resources/layout/activity_main.xml` so that the
    `<TextView/>` has `android:id="@+id/text_view"`:

        <TextView
          android:id="@+id/text_view"
          android:layout_width="wrap_content"                                     
          android:layout_height="wrap_content"                                    
          android:layout_centerInParent="true"                                    
          android:text="@string/app_text"                                         
        />

 3. Build!

        dotnet build -p:AndroidFragmentType=AndroidX.Fragment.App.Fragment \
	  -p:AndroidGenerateLayoutBindings=True \
	  -p:DefineConstants=__HAVE_ANDROIDX__

Expected results: no errors.

Actual results:

	…/LayoutBinding.cs(91,56): error CS0400: The type or namespace name 'Androidx' could not be found in the global namespace
	…/LayoutBinding.cs(91,135): error CS0400: The type or namespace name 'Androidx' could not be found in the global namespace

Update `LayoutBinding.cs` so that it uses the correct `AndroidX`
namespace.  Additionally, `FragmentManager` has been deprecated;
update `LayoutBinding.cs` to use `SupportFragmentManager`, which is
available on the `AndroidX.Fragment.App.FragmentActivity` class.

Finally, remove the use of a `static` field to share object state between
tasks. We have been on `msbuild` rather than `xbuild` on *nix for a while
now, so this workaround is no longer needed.

TODO: How is `__HAVE_ANDROIDX__` getting defined in #6893?

TODO: Add `LayoutCodeBehind.md` to [MicrosoftDocs/xamarin-docs][2].

[0]: https://docs.microsoft.com/en-us/xamarin/android/deploy-test/building-apps/build-items#androidfragmenttype
[1]: https://docs.microsoft.com/en-us/xamarin/android/deploy-test/building-apps/build-properties#androidgeneratelayoutbindings
[2]: https://github.com/MicrosoftDocs/xamarin-docs/tree/live/docs/android/deploy-test/building-apps
```

**Comment by jonpryor** [comment]
@grendello, @jpobst, @moljac : #6893 & #6894 introduce a question of "integration" between [Layout Codebehind](https://github.com/xamarin/xamarin-android/blob/16bbf6f0b0cd49f96f73197d28e8faf31aa9e499/Documentation/guides/LayoutCodeBehind.md) and the current AndroidX packages.

Part of that integration is already done, via [overriding `$(AndroidFragmentType)`](https://github.com/xamarin/AndroidX/blob/628cf049e0ef61e140cb17600df16ed801449814/source/AndroidXTargets.cshtml#L36).

However, the integration is incomplete, as `__HAVE_ANDROIDX__` is never set.

How should we complete the integration between Layout Codebehind & AndroidX?  Should the AndroidX package(s) add `$(DefineConstants)=$(DefineConstants);__HAVE_ANDROIDX__`?  Should some other form of integration be introduced?

**Comment by jonathanpeppers** [comment]
Do we have a general idea about the usage of this feature? If it's meant for primetime, it seems like it should be ported to a C# source generator at some point and be the default in the `dotnet new android` template.

**Comment by grendello** [comment]
@jonpryor IIRC, the idea was that the AndroidX packages will define the symbol, yes. 

@jonathanpeppers this was implemented a few years ago, it is as ready as it can be for prime time, but I've always wondered how actually useful for developers is this feature.  I don't think it got much use over the years

**Review by grendello — APPROVED** [review]

**Comment by dellis1972** [comment]
> Related: we should make `LayoutBinding.cs` nullable-reference-type aware; as-is, if `dotnet build` is used, using codebehind files results in lots of NRT-related warnings.
> 
> ```
> …/LayoutBinding.cs(22,90): warning CS8625: Cannot convert null literal to non-nullable reference type.
> …/LayoutBinding.cs(22,13): warning CS8618: Non-nullable field 'boundActivity' must contain a non-null value when exiting constructor. Consider declaring the field as nullable.
> …/LayoutBinding.cs(37,11): warning CS8600: Converting null literal or possible null value to non-nullable type.
> …/LayoutBinding.cs(79,56): warning CS8602: Dereference of a possibly null reference.
> …/LayoutBinding.cs(79,56): warning CS8603: Possible null reference return.
> ```

I will do this in a seperate PR.

**Pull request merged**